Header, navbar, and admin sidebar refactoring

davide-negretti opened the pull request:
## References
* Fixes #2656
* Fixes #2368 
* Fixes #2699
* Fixes #2178

## Description

### Purpose of this PR:
* change the structure of the header and the navbar in order to:
  * make customisation easier
  * improve responsivity
  * improve accessibility
* remove unnecessary styling from shared components in order to make customisation easier
* fix admin sidebar issues
  * sidebar width must be the same when vertical scrollbar is present
* improve overall accessibility of:
  * main navbar
  * menubar (login, language switch, ...)
  * admin sidebar
* decouple sidebar and header/navbar styling
* decouple custom and DSpace theme styling
* make color customisation easier

### List of changes
* Menu items (`*-menu-item` components)
  * minor changes have been made in order to better handle their styling from their parent components
* Header and navbar
  * the content of header, navbar, and header-navbar-wrapper has been rearranged
    * the navbar component only contains the main menu
    * the `<nav>` tag now contains navigation links only, as required by semantic HTML specification
* Accessibility
  * ARIA attributes and roles have been improved according to specification
* Sidebar
  * the way collapsing work has been changed in order to solve issues with vertical scrolling
  * other changes were required in order to decouple sidebar and navbar styling
* Theming
  * SCSS variables have been reorganised and unnecessary variables have been removed
  * new SCSS variables have been added in order to make color customisation easier
* host-window service
  * some utility functions for responsivity have been added
* Other changes
  * some unecessary code has been removed

### Accessibility

In order to test overall accessibility I used this Chrome extension that provides a visual representation of how screen readers render the page:
https://chromewebstore.google.com/detail/aria-devtools/dneemiigcbbgbdjlcdjjnianlikimpck

This is how the current layout is rendered:
[admin sidebar](https://github.com/DSpace/dspace-angular/assets/87638927/765df8b7-544c-4545-8345-3e96e627cf38)
[header](https://github.com/DSpace/dspace-angular/assets/87638927/f1852b1b-7e82-4acf-b709-697b6228b87f)

And this is how it is rendered with this PR:
![admin sidebar and header](https://github.com/DSpace/dspace-angular/assets/87638927/e59b6582-8ddd-48c0-b145-49c758e99e1b)

I also checked for errors with this Chrome extensions and there are none:
https://chromewebstore.google.com/detail/wave-evaluation-tool/jbbplnpkjmmeebjpijfedlgcdilocofh


Note that `menu` and `menuitem` are now being used, and dropdown menus are better handled.

### Instruction for reviewers
* Check that both _dspace_ and _custom_ theme work fine in both mobile and desktop mode
* Check that CSS variables such as theme colors work fine
* Check that header, navbar, and sidebar are fully accessible

## Future improvements
These components still require some changes, for which I'm going to open a new issue:
* user-menu component
  * should only contain a list of menu items
  * toggler button is used in desktop mode only, and it should be moved to the parent component
  * the outer wrapper should be moved to the parent components as well
  * menu items should be <li> elements in order to improve the collapsible navbar
  * it should be made more accessible through ARIA attributes
  * Bootstrap classes should be removed as their properties are being overridden
* dropdown menus
  * they should be made more accessible through ARIA attributes
  * Bootstrap classes should be removed as their properties are being overridden

## Checklist

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR passes [ESLint](https://eslint.org/) validation using `yarn lint`
- [x] My PR doesn't introduce circular dependencies (verified via `yarn check-circ-deps`)
- [x] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new libraries/dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR includes new features or configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

tdonohue commented:
@davide-negretti : Even though this is still a "Draft", I gave this a very brief test just now in production mode (`yarn build:prod; yarn serve:ssr`), just to see how it's working.  The basics seem to work. 

However, when I login, I receive a number of server-side rendering (SSR) errors which imply that SSR isn't working anymore:
```
ERROR RuntimeError: NG03402
    at triggerTransitionsFailed (C:\dspace-angular\dist\server\main.js:7:4969532)
    at TransitionAnimationEngine.reportError (C:\dspace-angular\dist\server\main.js:7:5048893)
    at TransitionAnimationEngine._flushAnimations (C:\dspace-angular\dist\server\main.js:7:5053458)
    at TransitionAnimationEngine.flush (C:\dspace-angular\dist\server\main.js:7:5048266)
    at InjectableAnimationEngine.flush (C:\dspace-angular\dist\server\main.js:7:5069338)
    at C:\dspace-angular\dist\server\main.js:7:5605397
    at _ZoneDelegate.invoke (C:\dspace-angular\dist\server\main.js:7:2179478)
    at Zone2.run (C:\dspace-angular\dist\server\main.js:7:2172289)
    at NgZone2.runOutsideAngular (C:\dspace-angular\dist\server\main.js:7:3215596)
    at AnimationRendererFactory.end (C:\dspace-angular\dist\server\main.js:7:5605336) {
  code: 3402
}
Error in server-side rendering (SSR)
Error details :  RuntimeError: NG03402
    at triggerTransitionsFailed (C:\dspace-angular\dist\server\main.js:7:4969532)
    at TransitionAnimationEngine.reportError (C:\dspace-angular\dist\server\main.js:7:5048893)
    at TransitionAnimationEngine._flushAnimations (C:\dspace-angular\dist\server\main.js:7:5053458)
    at TransitionAnimationEngine.flush (C:\dspace-angular\dist\server\main.js:7:5048266)
    at InjectableAnimationEngine.flush (C:\dspace-angular\dist\server\main.js:7:5069338)
    at InjectableAnimationEngine.ngOnDestroy (C:\dspace-angular\dist\server\main.js:7:5609542)
    at R3Injector.destroy (C:\dspace-angular\dist\server\main.js:7:2954388)
    at NgModuleRef.destroy (C:\dspace-angular\dist\server\main.js:7:3139781)
    at C:\dspace-angular\dist\server\main.js:7:3229024
    at Array.forEach (<anonymous>) {
  code: 3402
}
Falling back to serving direct client-side rendering (CSR).
```

davide-negretti commented:
@tdonohue this is now complete and can be tested. This PR is going to conflict with #2704, however I already handled all ARIA attributes

davide-negretti commented:
**Please note that I squashed some commits and force-pushed this branch**

tdonohue reviewed (commented):
@davide-negretti : Thanks! I gave this a test today.  I haven't yet done a full code review yet.

First, the good news is that I can verify that this fixes all three of these accessibility issues: #2368, #2699 and #2178

However, I'm still seeing several bugs:
1. I'm still seeing the [previously reported SSR errors](https://github.com/DSpace/dspace-angular/pull/2676#issuecomment-1843640938) whenever I login.  These errors no longer appear in my browser's DevTools console.  However, they still appear in the SSR output on the commandline in the window where I ran `yarn serve:ssr`.   So, we still seem to have SSR issues with the Transitions.
    * This SSR issue also may be causing the e2e test failures in this PR, as the e2e tests are reporting that same SSR bug over and over again.
2. There are minor style issues with the Admin Sidebar.  Here's a screenshot which compares the demo site (on the left) with the look of this PR (on the right).  The primary issue is that submenus are not indented, which causes the highlight around the menu (if you tab between menus) to be cutoff (see the highlight around "Community" in the menu below).  The spacing between menu items is also much smaller (not sure if that was intentional or not)
![menu-comparison](https://github.com/DSpace/dspace-angular/assets/483997/ff702cb5-c3a5-4db1-aa4e-3c002b43f3de)
3. Finally, the Admin Sidebar shifts slightly whenever you open or close the "New" menu.

Overall, this is good work & it appears to fix the accessibility issues.  But, there are still some bugs to fix.

davide-negretti commented:
@tdonohue the behavior of the sidebar when you open any submenu (causing the vertical scrollbar to appear) is intentional. Previously the vertical scrollbar caused the sidebar to become larger and sometimes it was covering the page content, although it's only 20 px larger. Now the width of the sidebar don't change anymore but after many attempts I found that the best way to keep the width constant (and the template code simple) was to slightly shift the menu items on the left, however this should not cause any issues as I reserved the needed space on the left.

I though I had already solved the SSR issue, which seemed to be caused by a wildcard, I will check it again. 

I will also check for the other layout issues, they are not intentional and I can fix them.

davide-negretti commented:
@tdonohue the e2e test `collection-statistics.cy.ts` is failing on main branch as well

![Schermata del 2023-12-15 16-58-17](https://github.com/DSpace/dspace-angular/assets/87638927/354c3a6d-1e29-401b-9b2e-fec4ce757e0f)

It throwns an error because it cannot find the "Statistics" navbar item at `/statistics/collections/282164f5-d325-4740-8dd1-fa4d6d3e7200`, and it is also missing on demo environment

tdonohue commented:
@davide-negretti : It's passing for me in my local environment on `main` and also passing in the last build of `main` (as of yesterday) in GItHub CI: https://github.com/DSpace/dspace-angular/actions/runs/7211583776

I will note though that there *are* some e2e tests which have random failures like this.  The various statistics e2e tests are all ones that I've seen have random failures (even in GitHub CI).   So, it is possible that these e2e failures are random in this PR... if you cannot find any other cause.  But, my suspicion is that the SSR errors may be having a side-effect of making these statistics pages _fail more frequently_.   So, if the SSR errors are fixed, it's possible these e2e failures will go away.  If they don't, I can see if I can dig deeper as to what the failure may be.

davide-negretti commented:
@tdonohue I fixed the issues with the animation in SSR. However the e2e tests keep failing on both this branch and main branch when I run them locally, and if I check the behaviour of the demo environment it does not match the expectations either:

![image](https://github.com/DSpace/dspace-angular/assets/87638927/11f520c7-fd49-44e9-931f-1ccfb4e70cd8)

tdonohue commented on src/app/shared/menu/menu-item/link-menu-item.component.html:
```diff
@@ -1,8 +1,6 @@
+<a class="ds-menu-item" role="menuitem"
    [ngClass]="{ 'disabled': !hasLink || item.disabled }"
    [attr.aria-disabled]="!hasLink || item.disabled"
-   [title]="item.text | translate"
```
@davide-negretti : I think this is why the e2e tests are failing in this PR.  If you look at the error in GitHub CI it says this:
> AssertionError: Timed out retrying after 4000ms: Expected to find element: `ds-navbar ds-link-menu-item a[title="Statistics"]`, but never found it.

These links no longer have a "title" text to match against.  So, the e2e tests for statistics need to be updated.  For example here: https://github.com/DSpace/dspace-angular/blob/main/cypress/e2e/collection-statistics.cy.ts#L9

I suspect the reason these e2e tests are not successful for you is because they REQUIRE the exact data from our Demo Entities Data set: https://github.com/DSpace-Labs/AIP-Files/releases/tag/demo-entities-data  If you have different data, or an expected Item/Collection/Community UUID is missing, then the e2e tests will fail.  While the demo site at https://demo.dspace.org does start with this data set, it's constantly being changed by people using the site...and those changes can result in e2e test failures.

davide-negretti replied:
@tdonohue thank you, I had to remove the title to fix an accessibility issues (the title is the same as the text and it is considered redundant) but I was focusing on the class selectors. I will fix the query accordingly

davide-negretti commented:
@tdonohue all tests are now passing

davide-negretti commented:
@tdonohue many conflicts were detected on this PR two days ago, as other PRs that fix some existing issues I also fixed have been recently merged. I merged `main` branch into my branch and in this way may changes in these PRs went lost. I'm trying to track these changes but I won't be able to keep many of them because they affect code I have refactored.

Also, now many e2e tests are failing, I'm trying to find the cause and I will update this PR as soon as possible.

tdonohue commented:
@davide-negretti : My apologies.  I'm trying to get Accessibility fixes merged as soon as I can, and it seems that I didn't realize merging #2711 or #2587 (those are the two accessibility PRs merged in the last few days) would negatively impact this one.

If you can get this updated one last time, I'll put it as "high priority" on my list to get merged.  Again, apologies...it can be difficult to tell which PRs will impact others prior to merging.

davide-negretti commented:
@tdonohue no problem, I just wanted to warn you about this problem in the case there were other pending PRs that affect the header and sidebar. I'm currently trying to fix e2e tests, there is one that is ramdomly failing due to timeout:

![image](https://github.com/DSpace/dspace-angular/assets/87638927/97ded9d7-0992-4b6f-8f62-3dd3e6834e2e)

![image](https://github.com/DSpace/dspace-angular/assets/87638927/edfdbd54-1081-48d4-b4ec-c366bfa17791)


![image](https://github.com/DSpace/dspace-angular/assets/87638927/96bb42e6-55c2-40b7-907c-38c11ec01f12)

tdonohue commented:
@davide-negretti : Don't worry as much about random-ish failures.  If they are truly random, then we can get this to pass & then fix the random failures in a future PR.  So, just concentrate on fixing the e2e tests that fail all the time.  Chances are they should all be something small like an Accessibility failure, or maybe you removed an HTML tag or attribute which the e2e test was depending on.

tdonohue reviewed (changes requested):
@davide-negretti : I've re-tested and re-reviewed this today.  Overall, the code looks good. I've verified it has no negative impact on things like theme overrides (at least I haven't found any side effects).

All the previously reported issues are fixed.  I've only found a *single, minor* new bug. 

* In the admin sidebar, you can no longer expand/contract sections using the keyboard.  To reproduce do the following:
    * Login as an admin
    * After login, use the "Tab" key on the keyboard to access the admin sidebar.
    * You should be able to tab to the various menu options successfully.
    * However, if you attempt to open a subsection (e.g. "Edit") by pressing "Enter", the section will open and then close again quickly.  There's no way to access the submenu options.

This is a minor issue, but it seems to be caused by this PR because it works fine on the demo site at https://demo.dspace.org/

I also looked at the current e2e failure & it appears that it may be a faulty e2e test.  I *cannot* reproduce it locally, but from the recorded video (in Github Actions there is a video recorded in the "Summary" whenever tests fail in GitHub) it appears a checkbox is not staying checked.  I think we may want to temporarily comment out the failing part of that test until I can investigate why it is failing.  I don't believe it to be related to this PR.  So, I'd recommend "fixing" it by commenting out line 48 of `collection-edit.cy.ts`: https://github.com/DSpace/dspace-angular/blob/main/cypress/e2e/collection-edit.cy.ts#L48

Beyond those two small things, this is looking great & nearly ready to merge.

davide-negretti commented:
@tdonohue I fixed the issue with expandable sections and I disabled the faulty test

tdonohue reviewed (approved):
👍 Thanks @davide-negretti !  This now looks good to me.  Everything seems to work the same as before, and I can confirm that the accessibility issues and duplicate elements in DOM issues appear to be fixed.

Because of the scale of this change, I would like to see if I can find a second set of eyes on this.  @artlowel or @alexandrevryghem : Would either of you be available to give this a quick look & test?  It's a major code refactor, but seems necessary to fix the duplicated DOM elements and accessibility issues noted.

artlowel commented on src/app/header/header.component.scss:
```diff
@@ -1,31 +1,30 @@
-  line-height: 1.5;
-}
+  .navbar-toggler .navbar-toggler-icon {
+    background-image: none !important;
```
I assume the `!important` is here to counter one from bootstrap, so that would be a legitimate case to use it (as that would be the only way to override another `!important`)

davide-negretti replied:
Yes, but this can actually be removed as I only needed the width and height of the `.navbar-toggler-icon` Bootstrap class.

Since these values are required only in the _base_ theme I'm going to hardcode them in the CSS files, as defining a CSS variable that is not used in the _dspace_ theme may be misleading

artlowel commented on src/app/header/header.component.scss:
```diff
@@ -1,31 +1,30 @@
-}
+  .navbar-toggler .navbar-toggler-icon {
+    background-image: none !important;
+    line-height: 1.5;
```
This should be (derived from) a variable

davide-negretti replied:
See comment above

https://github.com/DSpace/dspace-angular/pull/2676#discussion_r1475977391

artlowel commented on src/app/navbar/navbar-section/navbar-section.component.scss:
```diff
@@ -1,5 +1,5 @@
-  display: flex;
-  align-items: center;
-  height: 100%;
+:host {
+  li.ds-menu-item-wrapper {
+
+  }
```
Looks like this can be removed

davide-negretti replied:
I left it to make clear that this is the correct way to customise the items. Should I remove it anyway or should I add a comment to explain that?

artlowel commented on src/themes/dspace/app/header-nav-wrapper/header-navbar-wrapper.component.scss:
```diff
@@ -1,7 +1,91 @@
+        height: 100%;
+      }
+      .ds-menu-item {
+        padding: 0.5rem;
```
This should be (derived from) a variable. I also get the impression that dropdown menu options had more padding-left and right before this PR

artlowel commented on src/app/navbar/navbar.component.scss:
```diff
@@ -1,48 +1,64 @@
+    .ds-menu-item {
+      display: block;
+      color: var(--ds-navbar-link-color);
+      padding: 0.5rem 0;
```
This should be (derived from) a variable

tdonohue replied:
It looks like @artlowel 's feedback on this line wasn't resolved or answered.  I think he was asking for the `padding` to be derived from a variable

davide-negretti replied:
@tdonohue @artlowel I'm sorry, I though I already replied to this but I cannot see my answer anymore. In this case I chose not to use a variable because it would affect only the _base_ theme, and not the _dspace_ theme, and in my opinion having a global variable that does not change anything on the default _dspace_ theme may be confusing

artlowel commented on src/app/shared/auth-nav-menu/user-menu/user-menu.component.scss:
```diff
@@ -0,0 +1,12 @@
+
+  .ds-menu-item {
+    display: inline-block;
+    padding: 0.25rem 0;
```
This should be (derived from) a variable

artlowel commented on src/styles/_bootstrap_variables.scss:
```diff
@@ -20,48 +23,80 @@ $h4-font-size: $font-size-base * 1.25 !default;
+$ds-admin-sidebar-collapsible-element-width: 250px !default;
+$ds-admin-sidebar-total-width: $ds-admin-sidebar-fixed-element-width + $ds-admin-sidebar-collapsible-element-width; // do not change/override
+
+$ds-primary-bg: #{lighten($primary, 30%)} !default;
+$ds-secondary-bg: #{lighten($secondary, 30%)} !default;
+$ds-success-bg: #{lighten($success, 30%)} !default;
+$ds-info-bg: #{lighten($info, 30%)} !default;
+$ds-warning-bg: #{lighten($warning, 30%)} !default;
+$ds-danger-bg: #{lighten($danger, 30%)} !default;
+
```
Why have `ds` variants of these the bootstrap variables? why not simply override the original bg color variables? The way it is now, if a bootstrap component uses one of these bg color variables it will be the original bootstrap color, and when a dspace component does it it will be this lighter variant

tdonohue replied:
@davide-negretti : You appear to be referencing a commit from DSpace-CRIS.  These `$ds-*` variables do NOT exist in the DSpace codebase: https://github.com/DSpace/dspace-angular/blob/main/src/styles/_bootstrap_variables.scss

Instead, it appears the variables have a different name in DSpace: https://github.com/DSpace/dspace-angular/blob/main/src/styles/_bootstrap_variables.scss#L62-L67

So, maybe this means this PR should be updated to use the new name?

tdonohue replied:
Update: oh, I see... you renamed them and change their default settings a bit. 
For instance:
```
$primary-bg: #{lighten(map-get($theme-colors, primary), 30%)} !default;
```
Became...
```
$ds-primary-bg: #{lighten($primary, 30%)} !default;
```

davide-negretti replied:
@tdonohue exactly, I linked the commit from 4science repository but it has been made by Atmire on main branch

My intent was to make clear that these are custom DSpace variables that don't exist in Bootstrap, as the previous names may have suggested

artlowel replied:
Fair enough. Then I suggest that we simply remove the mapping to --bs css variables below to avoid that confusion. That doesn't necessarily have to happen in this PR since it doesn't add them.

davide-negretti replied:
@artlowel I agree, however some of these `--bs` CSS variables are being used so they cannot be deleted. I suggest to replace their prefix with `--ds` and move them together with other `--ds` variables.

artlowel replied:
Sounds good!

artlowel commented on src/styles/_bootstrap_variables_mapping.scss:
```diff
@@ -29,17 +29,17 @@
-  --bs-secondary-bg: #{$secondary-bg};
+  --bs-secondary-bg: #{$ds-secondary-bg};
   --bs-success: #{$success};
-  --bs-success-bg: #{$success-bg};
+  --bs-success-bg: #{$ds-success-bg};
   --bs-info: #{$info};
-  --bs-info-bg: #{$info-bg};
+  --bs-info-bg: #{$ds-info-bg};
   --bs-warning: #{$warning};
-  --bs-warning-bg: #{$warning-bg};
+  --bs-warning-bg: #{$ds-warning-bg};
   --bs-danger: #{$danger};
-  --bs-danger-bg: #{$danger-bg};
+  --bs-danger-bg: #{$ds-danger-bg};
```
If we _do_ end up needing dedicated ds versions of these bootstrap background colors, I still wouldn't override the bootstrap css variables with them. The idea behind this mapping is that it is exactly what's in the sass variables.

artlowel commented on src/themes/dspace/app/header-nav-wrapper/header-navbar-wrapper.component.scss:
```diff
@@ -1,7 +1,91 @@
+
+        // mobile menu
+        .ds-menu-item {
+          padding: 0.25rem 0;
```
This should be (derived from) a variable

artlowel commented on src/themes/dspace/app/header/header.component.scss:
```diff
@@ -1,26 +1,19 @@
-  @media screen and (max-width: map-get($grid-breakpoints, md)) {
-    height: var(--ds-header-logo-height-xs);
+  img#header-logo {
+    height: 40px;
```
We already have a variable for this: `var(--ds-header-logo-height)`

artlowel reviewed (changes requested):
Thanks @davide-negretti!

When you open a menu option and that causes a scrollbar to appear, the current sidebar on sandbox.dspace.org gets wider to account for this. In this PR all options jump to the left when the scrollbar appears

davide-negretti commented:
> When you open a menu option and that causes a scrollbar to appear, the current sidebar on sandbox.dspace.org gets wider to account for this. In this PR all options jump to the left when the scrollbar appears.

@artlowel this is a design choice I explained here:

https://github.com/DSpace/dspace-angular/pull/2676#issuecomment-1856601385

If it is not good I can revert it

artlowel commented:
@davide-negretti I'm sorry, I missed that comment. This occurs only in pinned mode I assume? Perhaps we can fix that by giving the sidebar 20px of (extra) margin-right when there's no scrollbar, and removing that when there is. That way the sidebar's width stays consistent, but to the user it will look like the scrollbar is added on the outside

davide-negretti commented:
@artlowel I already considered adding a left margin or padding, but we cannot add it directly to the `inner-wrapper` of the root component otherwise there would be a gap between full width elements (like the header) and the sidebar. The only solution would be to add an internal left padding to each direct child of the `inner-wrapper`, which may be an issue with more complex layouts. The best solution in my opinion would be to add an internal right padding to the sidebar, but I haven't been able to do that without breaking it.

davide-negretti commented:
@artlowel @tdonohue I completed all the changes that I marked as resolved, I'll wait for clarifications for the remaining ones

davide-negretti commented:
Merge conflicts solved

davide-negretti commented:
@tdonohue the issue with the logout button is caused by the Bootstrap `sr-only` class, whose implementation is quite a workaround

This happens when the header runs out of horizontal space

The current behaviour of the header is this:

![image](https://github.com/DSpace/dspace-angular/assets/87638927/046b1449-f723-4f2b-b328-ea202d26c883)

With my refactoring logo and buttons remain on the same line. I suggest to keep it in this way, and to hide the logo below 360px (or to replace it with a small `fa-house` icon):

![image](https://github.com/DSpace/dspace-angular/assets/87638927/50064bac-02a1-4d4e-b635-ee29bc5327fa)

This in my opinion looks better, and it also fixes the issue with `sr-only`

I could

artlowel reviewed (changes requested):
Thanks @davide-negretti!

>@artlowel I already considered adding a left margin or padding, but we cannot add it directly to the inner-wrapper of the root component otherwise there would be a gap between full width elements (like the header) and the sidebar. The only solution would be to add an internal left padding to each direct child of the inner-wrapper, which may be an issue with more complex layouts. The best solution in my opinion would be to add an internal right padding to the sidebar, but I haven't been able to do that without breaking it.

Ok, if there's no good way to do it I'm fine keeping it as is

For the horizontal space in the header issue: I think most institutions won't be happy if their logo is removed entirely. SO I'd go for the option where the menu slides _under_ it. However that's not what I'm seeing with the PR as it is. I'm seeing the same bug @tdonohue reported, and if you go even smaller you get a horizontal scrollbar.

I reviewed again, and I noticed that the public menu in this PR is no longer valid HTML:
```html
<ul>
    <ds-navbar-section>
        <li>...</li>
    </ds-navbar-section>
    <ds-themed-expandable-navbar-section>
        <ds-expandable-navbar-section>
            <li>...</li>
        </ds-expandable-navbar-section>
    </ds-themed-expandable-navbar-section>
</ul>
```

`<ul>`s are only allowed to contain `<li>` tags (or `<script>`s iirc). On the demo server this is the case. Iirc there should be a way to use an angular component with a default html tag (so e.g. `<ds-navbar-section>` would be rendered as an `<li>` itself) 

Another minor issue: the background when hovering over the Profile, MyDSpace etc options in the login dropdown have disappeared

davide-negretti commented:
@artlowel
- for the navbar issue I can use an attribute selector `[dsNavbarSection]` so that the `<ul>` will have only direct `<li dsNavbarSection>` descendants
- I will make the header height variable on mobile screens
- I will fix merge conflicts and the login dropdown issue

davide-negretti commented:
@artlowel 
- I replaced all <ul> and <li> with <div> in navbar and sidebar
- I made the header height variable on mobile screens. I did not fix the issue with `sr-only`, however this only occurs when the screen width is below 270px
- I restored the behaviour of links in the user dropdown menu

artlowel reviewed (changes requested):
Thanks @davide-negretti!

One more minor comment. Perhaps due to the switch from li's to divs, the spacing between options in the mobile menu seems to have decreased. It is especially noticeable, because the user menu above still uses li's

alexandrevryghem commented:
@davide-negretti: I think it would be best to remove the changes you made regarding to the `<ul>`/`<li>` structure replacement to nested `<div>`s, because this is not good for accessibility. Instead you could just replace [this `ng-container`](https://github.com/DSpace/dspace-angular/blob/523850bc45ed640d3331e07932d9eae8785d33c6/src/app/admin/admin-sidebar/admin-sidebar.component.html#L33) with an `li` and replace [this `<li>`](https://github.com/DSpace/dspace-angular/blob/523850bc45ed640d3331e07932d9eae8785d33c6/src/app/admin/admin-sidebar/admin-sidebar-section/admin-sidebar-section.component.html#L1) & [this `<li>`](https://github.com/DSpace/dspace-angular/blob/523850bc45ed640d3331e07932d9eae8785d33c6/src/app/admin/admin-sidebar/expandable-admin-sidebar-section/expandable-admin-sidebar-section.component.html#L1) with a `<div>`.

davide-negretti commented:
@alexandrevryghem I already added to every menu the `menubar`, `menu` and `menuitem` ARIA roles, which are more appropriate - from the accessibility point of view - than unordered lists, which uses the `list` and `listitem` roles.

I carefully tested the page with a screen reader, which previously recognised the menus as generic lists with N items, and now correctly recognises them as navigation bars.

I also inspected the page with ARIA DevTools chrome extension. This is the current situation on demo.dspace.org:

![image](https://github.com/DSpace/dspace-angular/assets/87638927/a2b6ac49-5120-4cb0-a415-ce9f603f0d5f)

And this is with the correct roles:

![image](https://github.com/DSpace/dspace-angular/assets/87638927/7efef583-250a-4dc9-96d5-07f76cef205a)

davide-negretti commented:
@artlowel I restored the spacing between items

![image](https://github.com/DSpace/dspace-angular/assets/87638927/d358e96e-659a-4b82-8616-2a3c6ada741c)

artlowel reviewed (approved):
Thanks @davide-negretti!

tdonohue commented:
@davide-negretti : As I worried, this was unable to be ported automatically to 7.x.  If you feel this work could be ported to 7.x, then I'd appreciate it if you could find time to create a version of this PR against the `dspace-7_x` branch.  If you feel this work is not easily possible to port to 7.x, then we can keep this as an 8.0-only accessibility fix.  

Please let me know what approach you feel is best.

davide-negretti commented:
@tdonohue I'm going to try to port it manually, I'll let youknow if the merge conflicts can be easily handled

davide-negretti commented:
@tdonohue
* I opened the PR for the porting: #2858 
* I made a minor fix for this PR (already included in the porting PR): #2859
